enhancements/machine-api/out-of-tree-provider-support.md

```diff
@@ -0,0 +1,553 @@
+
+*Note: metrics state is a subject to change, and we will have influence upon adding back ones required by OpenShift before transition to out-of-tree*
+
+### Disaster Recovery
+
+In the original `KCM` implementation, where the `CCM` is currently hosted, is using a [semi-automated](https://docs.openshift.com/container-platform/4.5/backup_and_restore/disaster_recovery/scenario-2-restoring-cluster-state.html) approach to disaster recovery, introducing the `forceRedeploymentReason` option in the managed CR, triggering the redeployment of the static pods, running the `KCM` code.
+
+`CCM` approach is simpler, as we run the `CCCMO` which manages the `Deployment`. Re-populating etcd will revert the old image, and the previous version of the `CCCMO` code will re-apply and restart the old `CCM` configuration.
+
+### Upgrade/Downgrade strategy
+
+A set of `cloud-controller-manager` pods will be running under a `Deployment` resource, provisioned by the operator. The leader-election will preserve the leadership during the cluster updates over cloud-specific loops.
+
+Upgrades from previous versions of OpenShift will look like:
+
+#### Post 4.7 (in-tree to out-of-tree per provider)
+
+Each provider will be released independently as and when the providers are deemed stable enough.
+
+Upgrade:
+
+- New `cluster-kube-apiserver-operator` and `cluster-kube-controller-manager-operator` version updates the `kube-apiserver` and the `KCM` and removes the `--cloud-provider` flag. `cluster-kube-controller-manager-operator` in addition sets the `--external-cloud-volume-plugin=<some-platform>` flag on the `KCM` pods to ensure in-tree storage preservation.
+- `cluster-cloud-controller-manager-operator` starts and creates out-of-tree `CCM` resources for the provider.
+- `machine-config-operator` restart updates the `kubelet` with `--cloud-provider=external` option. At this point the `CCM` is running out-of-tree
+
+
+Downgrade:
+
+- `cluster-cloud-controller-manager-operator` de-provisions CCM resources for any platform.
+- `cluster-kube-apiserver-operator` and `cluster-kube-controller-manager-operator` restart updates the `kube-apiserver` and the `kube-controller-manager` with the `--cloud-provider=<some-platform>` option.
+- `cluster-kube-controller-manager-operator` removes the `--external-cloud-volume-plugin=<some-platform>` flag from the `KCM` pods as `--cloud-provider=<some-platform>` superseeds it.
+- `machine-config-operator` restart applies an old machine configuration, and cause `kubelet` to restart with `--cloud-provider=<some-platform>` option.
+
+#### Providers, already running out-of-tree CCM (4.8+)
+
+Upgrade:
+
+- `cluster-cloud-controller-manager-operator` updates the state of resources for any in-tree provider defaulting to this since previous releases.
+
+Downgrade:
+
+- `cluster-cloud-controller-manager-operator` downgrades the state of resources for any in-tree provider defaulting to this since previous releases.
+
+### Version Skew Strategy
```

Comment:
Current eus->eus upgrade strategy expectes that some kubelet updates are skipped. I.e, kcm may be 4.10, while kubelet is 4.8 (numbers are purely illustrative). Can it break anything, esp. when the feature is GA in 4.10 and still optional / disabled by default in 4.8?

Reply:
Not unless we integrate migration functionality in 4.8. That's why starting earlier is important. The gap may result in core providers not reporting necessary conditions in time, so the migration may not initiate.

Reply:
Is this still true? I was of the impression we had decided EUS would mean 4 baby upgrades to get between EUS versions? IE we can't skip ahead?